builders_guide.md

```diff
@@ -0,0 +1,551 @@
+
+        const planetAfterUpgrade = await world.contracts.core.planets(silverMinePlanetId);
+        const newPopulationCap = planetAfterUpgrade.populationCap;
+        const newPopulationGrowth = planetAfterUpgrade.populationGrowth;
+        const newSilver = planetAfterUpgrade.silver.toNumber();
+
+        expect(newSilver).to.equal(initialSilver - 0.2 * silverCap);
+        expect(initialPopulationCap).to.be.below(newPopulationCap);
+        expect(initialPopulationGrowth).to.be.below(newPopulationGrowth);
+      });
+  ```
+</details>
+
+## 3.4 Run the new Hardhat test
+
+- `yarn workspace eth hardhat test test/DFCustomUpgrade.test.ts`
+- You should see this output
+    
+    ![Custom upgrade test output](img/custom_upgrade_final.png)
+    
+
+## 3.5 Run all Hardhat tests
+
+- `yarn workspace eth test`
+- You will see an error in `DFUpgrade.test.ts`:
+    
+    > AssertionError: Expected transaction to be reverted with Can only upgrade regular planets, but other exception was thrown: Error: VM Exception while processing transaction: reverted with reason string 'Can only upgrade allowed planet types’
+    > 
+    - How would you fix this?
+        - (Hint: see the Solution to 3.3.1)
+- Fix this error and run `yarn workspace eth test` again and all test should pass.
+
+The upgrade logic is now working in the tests, but we need to make these changes accessible to players using the game client as well.
+
+## 3.6 Add `UPGRADEABLE_PLANETS` config to the client
+
+- This allows us to reference the value of `UPGRADEABLE_PLANETS` anywhere in the client.
+
+### 3.6.1 Add `UPGRADEABLE_PLANETS` type
+
+- The Dark Forest client is a TypeScript project. That means that for every new variable we add, (like `UPGRADEABLE_PLANETS`), we need to add its type as well.
+- In `client/src/_types/darkforest/api/ContractsAPITypes.ts:`
+    
+    ```diff
+    export interface ContractConstants {
+       ...
+    +  UPGRADEABLE_PLANETS: boolean[];
+    }
+    ```
+    
+
+### 3.6.2 Add `UPGRADEABLE_PLANETS` value
+
+- In `client/src/Backend/GameLogic/ContractsAPI.ts`
+    
+    ```typescript
+    // ~ Line 800
+    {
+    ...
+    UPGRADEABLE_PLANETS
+    } = await this.makeCall(this.coreContract.gameConstants);
+    
+    // ~ Line 880
+    
+    const constants: ContractConstants = {
+    ...
+    UPGRADEABLE_PLANETS
+    }
+    ```
+    
+
+## 3.7 Show user which planets are upgradeable
+
+- In `client/src/Frontend/Panes/UpgradeDetailsPane.ts`:
+    - Try to change this logic yourself to allow only the planets marked as true in `UPGRADEABLE_PLANETS` variable to be upgraded in the client.
+    - Hint: This is very similar to the change we made to `DarkForestPlanet.sol` but in TypeScript.
+    - Also, to get access to the `UPGRADEABLE_PLANETS` value, we need to access the GameManager object. The React [Frontend](client/src/Frontend) doesn’t have direct access to the GameManager, but it does have access to the GameUiManager, which can fetch the GameManager.
```

Comment:
```suggestion
    - Also, to get access to the `UPGRADEABLE_PLANETS` value, we need to access the GameManager object. The React [Frontend](https://github.com/darkforest-eth/client/tree/master/src/Frontend) doesn’t have direct access to the GameManager, but it does have access to the GameUiManager, which can fetch the GameManager.
```
It seems like we can't link into the submodules from github UI, so let's direct to the repo